### PR TITLE
Crypto: BLS-based SPoCK scheme 

### PR DESCRIPTION
**Pull request by tarakby**
This PR replaces the temporary naive implementation of the SPoCK API by the BLS-based implementation.

### CODE REVIEW

File: crypto/bls12381_utils.c [hunk 1]
```diff
@@ -568,6 +568,73 @@ void ep2_sum_vector(ep2_t jointy, ep2_st* y, const int len){
...
+    // check s2 is on curve and in G1
+    if (check_membership_G1(elemsG1[1]) != VALID) // only enabled if MEMBERSHIP_CHECK==1
+        return INVALID; 
+
+    // elemsG2[1] = pk1
+    ep2_new(elemsG2[1]);
+    ep2_copy(elemsG2[1], (ep2_st*)pk1);
+
+    // elemsG2[0] = pk2
+    ep2_new(&elemsG2[0]);
```
Comment (AlexHentschel) [inline crypto/bls12381_utils.c:603]: here, you use the memory address `&elemsG2[0]` of point pk2 while above, you use the value:
https://github.com/onflow/flow-go/blob/de7518dbbeffc54db495dc842c73181ea3db0506/crypto/bls12381_utils.c#L599

is that intentional?

**Review by AlexHentschel — APPROVED** [review]
Looks good. Unfortunately, I don't have the background to really understand whether the formulas are correct.

File: crypto/bls12381_utils.c (hunk 1 above)
Comment (tarakby) [reply crypto/bls12381_utils.c:603]: Good catch! It doesn't matter with the Relic settings we're using now but it does matter if I change the settings.

File: crypto/bls.go [hunk 2]
```diff
@@ -20,11 +20,15 @@ package crypto
...
 //  - membership check in G2 is using a simple scalar multiplication with the group order.
 //  - multi-signature tools are defined in bls_multisg.go
+//  - SPoCK scheme based on BLS: verifies two signatures have been generated from the same message,
+//    that is unknown to the verifier.
 
 // future features:
 //  - membership checks G2 using Bowe's method (https://eprint.iacr.org/2019/814.pdf)
 //  - implement a G1/G2 swap (signatures on G2 and public keys on G1)
 
+// #cgo coverage CFLAGS: --coverage -fprofile-arcs -ftest-coverage -fprofile-instr-generate -fcoverage-mapping
```
Comment (tarakby) [inline crypto/bls.go:30]: to delete

File: crypto/spock.go [hunk 3]
```diff
@@ -10,54 +10,51 @@ package crypto
...
-// is only to facilitate the SPoCK integration in Flow. The implementation
-// passes the happy and unhappy path tests but does not satisfy any security
-// or correctness property.
-
-// SPOCKProve generates a spock proof for data under the private key sk.
+// SPOCKProve generates a spock poof for data under the private key sk.
 func SPOCKProve(sk PrivateKey, data []byte, kmac hash.Hasher) (Signature, error) {
 	if sk.Algorithm() != BLSBLS12381 {
-		return nil, errors.New("private key must be a BLS key")
+		return nil, errors.New("private key must be a BLS key.")
```
Comment (tarakby) [inline crypto/spock.go:21]: print the algo

File: crypto/spock.go [hunk 4]
```diff
@@ -10,54 +10,51 @@ package crypto
...
+	blsPk1, ok1 := pk1.(*PubKeyBLSBLS12381)
+	blsPk2, ok2 := pk2.(*PubKeyBLSBLS12381)
+	if !(ok1 && ok2) {
+		return false, errors.New("public keys must be BLS keys.")
 	}
-	return bytes.Equal(pk1.Encode()[:10], proof1[:10]) &&
-		bytes.Equal(pk2.Encode()[:10], proof2[:10]) &&
-		bytes.Equal(proof1[10:], proof2[10:]), nil
+
+	if len(proof1) != signatureLengthBLSBLS12381 || len(proof2) != signatureLengthBLSBLS12381 {
```
Comment (tarakby) [inline crypto/spock.go:49]: add a comment for not returning an error (also in ecdsa and bls verify)

File: crypto/bls12381_utils.c [hunk 5]
```diff
@@ -568,6 +568,73 @@ void ep2_sum_vector(ep2_t jointy, ep2_st* y, const int len){
...
+    // elemsG1[0] = s1
+    ep_new(elemsG1[0]);
+    if (ep_read_bin_compact(elemsG1[0], sig1, SIGNATURE_LEN) != RLC_OK) 
+        return INVALID;
+
+    // check s1 is on curve and in G1
+    if (check_membership_G1(elemsG1[0]) != VALID) // only enabled if MEMBERSHIP_CHECK==1
+        return INVALID;
+
+    // elemsG1[0] = s2
```
Comment (tarakby) [inline crypto/bls12381_utils.c:589]: upd comment

**Review by jordanschalm — APPROVED** [review]
Tarak walked us through changes on a call

**Review by ramtinms — APPROVED** [review]

**Review by zhangchiqing — APPROVED** [review]

**Pull request merged**